### PR TITLE
Ph2 TK Digitizer software updated for RandomNumberGenerator service

### PR DESCRIPTION
**Pull request by suchandradutta**
The usage of edm::RandomNumberGenerator reviewed following the comment in [1]. We are now generating random numbers in each event independently. The code is accordingly updated.
@makortel, @boudoul, @emiglior 

[1] https://github.com/cms-sw/cmssw/issues/22710#issuecomment-375337651

### CODE REVIEW

**Comment by boudoul** [comment]
Thanks @suchandradutta , Could you please edit the title mentioning that this is a fix in the random generator usage in the phase2 tracker digitizer  to be a bit more explicit ?  Also could you please backport to 93X ? Thank you again

**Comment by suchandradutta** [comment]
@boudoul yes I shall back port it for 93X once the PR is reviewed and accepted. I hope you agree.

**Comment by boudoul** [comment]
sure, thanks

File: SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizer.cc [hunk 1]
```diff
@@ -93,17 +93,14 @@ namespace cms
     mixMod.produces<edm::DetSetVector<Phase2TrackerDigi> >("Tracker").setBranchAlias(alias2);
     mixMod.produces<edm::DetSetVector<PixelDigiSimLink> >("Tracker").setBranchAlias(alias2);
 
+    // creating algorithm objects and pushing them into the map
+    algomap_[AlgorithmType::InnerPixel] = std::unique_ptr<Phase2TrackerDigitizerAlgorithm>(new PixelDigitizerAlgorithm(iconfig_));
+    algomap_[AlgorithmType::PixelinPS]  = std::unique_ptr<Phase2TrackerDigitizerAlgorithm>(new PSPDigitizerAlgorithm(iconfig_));
+    algomap_[AlgorithmType::StripinPS]  = std::unique_ptr<Phase2TrackerDigitizerAlgorithm>(new PSSDigitizerAlgorithm(iconfig_));
+    algomap_[AlgorithmType::TwoStrip]   = std::unique_ptr<Phase2TrackerDigitizerAlgorithm>(new SSDigitizerAlgorithm(iconfig_));
```
Comment (makortel) [inline SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizer.cc]: Right-hand side could be written more shortly as `std::make_unique<PixelDigitizerAlgorithm>(iConfig);` etc.

On a more general note (possibly going beyond this PR)
* `algomap_` could be implemented as `std::vector` (consumes less memory, faster access)
* `iconfig_` could now be removed if `isOTreadoutAnalog` parameter would be read to member variable

File: SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizer.cc [hunk 2]
```diff
@@ -123,21 +120,14 @@ namespace cms
...
-    // creating algorithm objects and pushing them into the map
-    algomap_[AlgorithmType::InnerPixel] = std::unique_ptr<Phase2TrackerDigitizerAlgorithm>(new PixelDigitizerAlgorithm(iconfig_, (*rndEngine_)));
-    algomap_[AlgorithmType::PixelinPS]  = std::unique_ptr<Phase2TrackerDigitizerAlgorithm>(new PSPDigitizerAlgorithm(iconfig_, (*rndEngine_)));
-    algomap_[AlgorithmType::StripinPS]  = std::unique_ptr<Phase2TrackerDigitizerAlgorithm>(new PSSDigitizerAlgorithm(iconfig_, (*rndEngine_)));
-    algomap_[AlgorithmType::TwoStrip]   = std::unique_ptr<Phase2TrackerDigitizerAlgorithm>(new SSDigitizerAlgorithm(iconfig_, (*rndEngine_)));
   }
-
   void Phase2TrackerDigitizer::endLuminosityBlock(edm::LuminosityBlock const& lumi, edm::EventSetup const& iSetup) {
-    algomap_.clear();
+    edm::LogInfo("Phase2TrackerDigitizer") << "End Luminosity Block";
```
Comment (makortel) [inline SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizer.cc]: Please remove this function as it does nothing useful anymore.

File: SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizer.cc [hunk 3]
```diff
@@ -123,21 +120,14 @@ namespace cms
...
   }
+
   Phase2TrackerDigitizer::~Phase2TrackerDigitizer() {  
     edm::LogInfo("Phase2TrackerDigitizer") << "Destroying the Digitizer";
+    algomap_.clear();
```
Comment (makortel) [inline SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizer.cc]: Calling `clear()` is unnecessary (`algomap_`'s destructor will clean up all its contents automatically).

File: SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc [hunk 4]
```diff
@@ -152,14 +151,13 @@ Phase2TrackerDigitizerAlgorithm::Phase2TrackerDigitizerAlgorithm(const edm::Para
...
-  // Threshold smearing with gaussian distribution:
-  smearedThreshold_Endcap_(addThresholdSmearing ? new CLHEP::RandGaussQ(eng, theThresholdInE_Endcap , theThresholdSmearing_Endcap) : nullptr),
-  smearedThreshold_Barrel_(addThresholdSmearing ? new CLHEP::RandGaussQ(eng, theThresholdInE_Barrel , theThresholdSmearing_Barrel) : nullptr),
-  rengine_(&eng)
+  subdetEfficiencies_(conf_specific)
 {
+  flatDistribution_.reset(nullptr);
+  gaussDistribution_.reset(nullptr);
+  smearedThreshold_Endcap_.reset(nullptr);
+  smearedThreshold_Barrel_.reset(nullptr);
```
Comment (makortel) [inline SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc]: These are not needed, `unique_ptr` default is `nullptr`.

File: SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc [hunk 5]
```diff
@@ -726,6 +724,21 @@ void Phase2TrackerDigitizerAlgorithm::pixel_inefficiency(const SubdetEfficiencie
...
+void Phase2TrackerDigitizerAlgorithm::initializeEvent(CLHEP::HepRandomEngine* eng) {
+  if (addNoise || AddPixelInefficiency || fluctuateCharge || addThresholdSmearing) {
+    
+    flatDistribution_ = std::unique_ptr<CLHEP::RandFlat>(new CLHEP::RandFlat(*eng, 0., 1.));
+    gaussDistribution_ = std::unique_ptr<CLHEP::RandGaussQ>(new CLHEP::RandGaussQ(eng, 0., theReadoutNoise));
+  }
+  // Threshold smearing with gaussian distribution:
+  if (addThresholdSmearing) {
+    smearedThreshold_Endcap_ =  std::unique_ptr<CLHEP::RandGaussQ> (new CLHEP::RandGaussQ(eng, theThresholdInE_Endcap , theThresholdSmearing_Endcap));
+    smearedThreshold_Barrel_ = std::unique_ptr<CLHEP::RandGaussQ> (new CLHEP::RandGaussQ(eng, theThresholdInE_Barrel , theThresholdSmearing_Barrel));
```
Comment (makortel) [inline SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc]: `std::make_unique<T>()` could be used here as well (if these are still needed, ref. @civanch's comment https://github.com/cms-sw/cmssw/pull/22836#discussion_r179105152).

**Review by makortel — COMMENTED** [review]
@suchandradutta Thanks! I gave some suggestions for further improvements (all not necessarily for this PR).

**Comment by kpedro88** [comment]
@boudoul is there a specific need for this in 93X?

**Comment by boudoul** [comment]
hi @kpedro88 , I do not have any particularly need. However when I do see in the description of the issue #22710 (adressed by this PR) how not safe it is (with many details explained in this issue) and knowing that 93X is intensively used for MC prod, I would feel better if this fix is also propagated there.

**Comment by kpedro88** [comment]
Okay, I hadn't read #22710. At some point we plan to leave 9_3_X behind, so unless UPSG is going to stop current production to wait for this, I'm not sure if a backport will get merged. Well, we can still make the backport PR so we have it if needed.

**Comment by civanch** [comment]
Is there the same issue in run-2 Digi, or only Phase2?

**Comment by makortel** [comment]
@civanch No, this issue is specific to phase2.

**Comment by kpedro88** [comment]
please test

**Comment by kpedro88** [comment]
The errors are caused by this PR, but I had to run valgrind to find the underlying problem:
```
==3530184== Invalid free() / delete / delete[] / realloc()
==3530184==    at 0x402925A: operator delete(void*) (in /cvmfs/cms.cern.ch/slc6_amd64_gcc630/external/valgrind/3.13.0-omkpbe/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3530184==    by 0xC11F2A0: _M_release (shared_ptr_base.h:150)
==3530184==    by 0xC11F2A0: ~__shared_count (shared_ptr_base.h:662)
==3530184==    by 0xC11F2A0: ~__shared_ptr (shared_ptr_base.h:928)
==3530184==    by 0xC11F2A0: ~shared_ptr (shared_ptr.h:93)
==3530184==    by 0xC11F2A0: CLHEP::RandGauss::~RandGauss() (RandGauss.cc:50)
==3530184==    by 0xC121CB8: CLHEP::RandGaussQ::~RandGaussQ() (RandGaussQ.cc:27)
==3530184==    by 0x2AD32EEE: operator() (unique_ptr.h:76)
==3530184==    by 0x2AD32EEE: reset (unique_ptr.h:347)
==3530184==    by 0x2AD32EEE: operator= (unique_ptr.h:254)
==3530184==    by 0x2AD32EEE: Phase2TrackerDigitizerAlgorithm::initializeEvent(CLHEP::HepRandomEngine*) (Phase2TrackerDigitizerAlgorithm.cc:726)
==3530184==    by 0x2AD211CC: cms::Phase2TrackerDigitizer::initializeEvent(edm::Event const&, edm::EventSetup const&) (Phase2TrackerDigitizer.cc:175)
```

**Comment by suchandradutta** [comment]
@makortel and @civanch  can you kindly help me to understand the issue? Please check the updated code [1] where I tried to follow your earlier suggestion. I have probably misunderstood the suggestion.
Kindly check and point where I have gone wrong.

[1] https://github.com/cms-sw/cmssw/pull/22836/commits/a1a8849a60de2599723c5717859bf14bd9b7b12c

**Comment by makortel** [comment]
@kpedro88 Could you share the full output from the job with valgrind?

**Comment by makortel** [comment]
@suchandradutta Taking a look of source of `RandGauss` constructor, it looks to me that if it is given `CLHEP::RandomEngine *` (as you do), it assumes the ownership of the object
https://github.com/cms-externals/clhep/blob/9cb9cc259aed1fd0dff85f100a9842e2304f8d4b/Random/Random/RandGauss.icc#L22-L25

The ownership "transfer" can be avoided by passing a reference instead
https://github.com/cms-externals/clhep/blob/9cb9cc259aed1fd0dff85f100a9842e2304f8d4b/Random/Random/RandGauss.icc#L17-L20

You could e.g. de-reference the pointer in `initializeEvent()`, i.e.
```cpp
gaussDistribution_ = std::make_unique<CLHEP::RandGaussQ>(*eng, 0., theReadoutNoise);
```
or change interface of `initializeEvent()` to take `CLHEP::RandomEngine` by reference
```cpp
void Phase2TrackerDigitizerAlgorithm::initializeEvent(CLHEP::HepRandomEngine& eng) {
```

This problem does not occur with the current code as `CLHEP::RandomEngine` is passed by reference to `RandGaussQ` constructor
https://github.com/cms-sw/cmssw/blob/f248aeaafe5327cc17cfeb919d95c8d865e094f9/SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc#L61
https://github.com/cms-sw/cmssw/blob/f248aeaafe5327cc17cfeb919d95c8d865e094f9/SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc#L156-L160

**Comment by makortel** [comment]
Given the possibly dangerous interface of `RandGauss` etc, I wonder if it would be worth of the effort to add a check to e.g. static analyzer? I would imagine in CMSSW we should always pass `CLHEP::RandomEngine` by reference.
@Dr15Jones

**Comment by kpedro88** [comment]
@makortel the valgrind log is here on cmslpc:
```
/uscms_data/d3/pedrok/phase2/test/CMSSW_10_2_0_pre1/src/test/log_valgrind2.log
```

**Comment by makortel** [comment]
@kpedro88 Thanks but I don't have account there. Anyway, I believe I figured out the cause so I don't have immediate need for the log anymore.

**Comment by kpedro88** [comment]
Okay, copied it to `/afs/cern.ch/user/p/pedrok/public/log_valgrind2.log` in case you want to look.

File: SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizer.cc [hunk 6]
```diff
@@ -123,19 +120,11 @@ namespace cms
...
-    algomap_[AlgorithmType::TwoStrip]   = std::unique_ptr<Phase2TrackerDigitizerAlgorithm>(new SSDigitizerAlgorithm(iconfig_, (*rndEngine_)));
   }
-
   void Phase2TrackerDigitizer::endLuminosityBlock(edm::LuminosityBlock const& lumi, edm::EventSetup const& iSetup) {
-    algomap_.clear();
+    edm::LogInfo("Phase2TrackerDigitizer") << "End Luminosity Block";
   }
+
   Phase2TrackerDigitizer::~Phase2TrackerDigitizer() {  
     edm::LogInfo("Phase2TrackerDigitizer") << "Destroying the Digitizer";
```
Comment (civanch) [inline SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizer.cc]: the best advice do not print in a destructor except during debugging of a problem. After debugging better clean up destructors.

**Comment by civanch** [comment]
@suchandradutta , this is an obvious double deletion of random engine. This happens when RandGaussQ is deleted. I would check CLHEP 2.4.0.0 headers description where authors provide detailed instructions which constructors/interfaces are thread safe. If those where the engine is defined by reference - use it. Also not necessary to have many RandGauss objects - by default there is a constructor with mean 0., RMS 1., this one for me is the best.

**Comment by kpedro88** [comment]
please test

**Comment by kpedro88** [comment]
Changes throughout Phase 2 workflows due to difference in random number usage, consistent with statistical fluctuations

**Comment by kpedro88** [comment]
+1

**Comment by civanch** [comment]
+1

**Comment by fabiocos** [comment]
@suchandradutta @makortel as far as I can see this PR is complementary to #22874 and not conflicting, right?

**Comment by makortel** [comment]
@fabiocos Yes, they are independent.

**Comment by fabiocos** [comment]
+1

**Pull request merged**